rest-api/src/main/java/gov/cms/qpp/conversion/api/internal/pii/SpecPiiValidator.java

```diff
@@ -21,19 +21,31 @@ public SpecPiiValidator(CpcValidationInfoMap file) {
 	}
 
 	@Override
-	public void validateApmNpiCombination(Node node, NodeValidator validator) {
+	public void validateApmTinNpiCombination(Node node, NodeValidator validator) {
 		String apm = node.getValue(ClinicalDocumentDecoder.PRACTICE_ID);
 		List<String> npiList = Arrays.asList(
 			node.getValue(ClinicalDocumentDecoder.NATIONAL_PROVIDER_IDENTIFIER).split(","));
-
-		for (final String npi : npiList) {
-			CpcValidationInfo spec = file.getApmToSpec().get(npi);
-			if (spec == null) {
-				validator.addWarning(Detail.forErrorAndNode(ErrorCode.INCORRECT_API_NPI_COMBINATION, node));
-			} else {
-				if (spec.getApm() != null && !spec.getApm().equalsIgnoreCase(apm)) {
+		List<String> tinList = Arrays.asList(
+			node.getValue(ClinicalDocumentDecoder.TAX_PAYER_IDENTIFICATION_NUMBER).split(","));
+
+		List<CpcValidationInfo> tinNpiCombinationList = file.getApmTinNpiCombination().get(apm);
+		if (tinNpiCombinationList == null){
+			validator.addWarning(Detail.forErrorAndNode(ErrorCode.INCORRECT_API_NPI_COMBINATION, node));
+		} else {
+			for (int index = 0; index < npiList.size(); index++) {
```

Comment:
how large are these lists? if they're very big, let's cache the size

Reply:
I think they can get fairly large, are you referring to caching with Integer.valueOf()? Or is the a preferred way for you.

Reply:
@saquino0827  the complexity here can be reduced if we store the CpcValidationInfo using maps. I am responding to this from the question that @omnivion posted below about caching. 
The QrdaService is injected into the RestController (which means it is a singletone). Also the `gov.cms.qpp.conversion.api.services.QrdaServiceImpl#cpcValidationData` is eagerly cached (2 hours). That means everything is in memory. 
I do not know how the incoming data for validation or access parttern is, I am assuming that the flow is something like this  `apm -> tin -> npi` or `apm -> npi -> tin`
Example:
```
//assuming that the access pattern is apm > tin > npi
//CpcValidationInfoMap.java
class CpcValidationInfoMap {
    Map<String, Map<String, List<String>> apmTinNpiCombinations = .....;
    public convertJsonToMapOfLists()...... populate...correctly the map.... 
    
    public boolean isValidRelation(String apm, String tin, String npi) { 
        if (apmTinNpiCombinations[apm] == null) return false;
        if (apmTinNpiCombinations[apm][tin]== null) return false;
        return (apmTinNpiCombinations[apm][tin]).indexOf(npi) > -1;
   }
}
Then in TNAHelper -> tinNpiCombinationExists you can call isValidRelation method
```

Reply:
@bijujoseph the problem with this is that one tin can have multiple NPI's associated with it. So you actually cannot do a Tin -> NPI map because of a duplicate keys issue.

Reply:
I meant cache the value of list.size() so we're not calling it 1000s of times if it's a huge list

Reply:
👍